parser/parser_system1.c

```diff
@@ -50,6 +50,91 @@
 #define PV_NS_UTS	0x2
 #define PV_NS_IPC	0x4
 
+static int parse_disks(struct pv_state *s, char *value, int n)
+{
+	int tokc, size, ret = 1;
+	char *str = NULL;
+	jsmntok_t *tokv, *t;
+
+	if (jsmnutil_parse_json(value, &tokv, &tokc) < 0) {
+		pv_log(ERROR, "wrong format filter");
+		goto out;
+	}
+
+	size = jsmnutil_array_count(value, tokv);
+	if (size <= 0) {
+		pv_log(ERROR, "empty filter");
+		goto out;
+	}
+
+	t = tokv + 1;
+	while ((str = pv_json_array_get_one_str(value, &size, &t))) {
+		struct pv_disk *d;
+		char *tmp;
+		jsmntok_t *diskv;
+		int diskc;
+
+		if (jsmnutil_parse_json(str, &diskv, &diskc) <= 0) {
+			pv_log(ERROR, "Invalid disk entry");
+			continue;
+		}
+
+		d = pv_disk_add(s);
+		if (!d) {
+			free(diskv);
+			goto out;
+		}
+
+		d->name = pv_json_get_value(str, "name", diskv, diskc);
+		d->path = pv_json_get_value(str, "path", diskv, diskc);
+		d->options= pv_json_get_value(str, "options", diskv, diskc);
+		d->uuid= pv_json_get_value(str, "uuid", diskv, diskc);
+
+		tmp = pv_json_get_value(str, "type", diskv, diskc);
+		if (!strcmp(tmp, "directory"))
+			d->type = DISK_DIR;
+		else if(!strcmp(tmp, "dm-crypt-versatile"))
+			d->type = DISK_DM_CRYPT_VERSATILE;
+		else if(!strcmp(tmp, "dm-crypt-caam"))
+			d->type = DISK_DM_CRYPT_CAAM;
+		else if(!strcmp(tmp, "dm-crypt-dcp"))
+			d->type = DISK_DM_CRYPT_DCP;
+		else {
+			d->type = DISK_UNKNOWN;
+			goto out;
```

Comment:
shouldnt we go where we free tokv etc.?

Reply:
```
out:
        if (str)
                free(str);

        if (tokv)
                free(tokv);
```
out is already freeing it. Am I missing something?

Reply:
not sure if str?